expression/builtin_json.go

```diff
@@ -702,8 +703,115 @@ type jsonArrayAppendFunctionClass struct {
 	baseFunctionClass
 }
 
+type builtinJSONArrayAppendSig struct {
+	baseBuiltinFunc
+}
+
+func (c *jsonArrayAppendFunctionClass) verifyArgs(args []Expression) error {
+	if len(args) < 3 || (len(args)&1 != 1) {
+		return ErrIncorrectParameterCount.GenWithStackByArgs(c.funcName)
+	}
+	return nil
+}
+
 func (c *jsonArrayAppendFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
-	return nil, errFunctionNotExists.GenWithStackByArgs("FUNCTION", "JSON_ARRAY_APPEND")
+	if err := c.verifyArgs(args); err != nil {
+		return nil, err
+	}
+	argTps := make([]types.EvalType, 0, len(args))
+	argTps = append(argTps, types.ETJson)
+	for i := 1; i < len(args)-1; i += 2 {
+		argTps = append(argTps, types.ETString, types.ETJson)
+	}
+	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETJson, argTps...)
+	for i := 2; i < len(args); i += 2 {
+		DisableParseJSONFlag4Expr(args[i])
+	}
+	sig := &builtinJSONArrayAppendSig{bf}
+	sig.setPbCode(tipb.ScalarFuncSig_JsonArrayAppendSig)
+	return sig, nil
+}
+
+func (b *builtinJSONArrayAppendSig) Clone() builtinFunc {
+	newSig := &builtinJSONArrayAppendSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinJSONArrayAppendSig) checkError(err error) error {
+	if err == nil || b.ctx.GetSessionVars().StrictSQLMode {
+		return err
+	}
+	b.ctx.GetSessionVars().StmtCtx.AppendWarning(err)
+	return nil
+}
+
+func (b *builtinJSONArrayAppendSig) evalJSON(row chunk.Row) (res json.BinaryJSON, isNull bool, err error) {
+	res, isNull, err = b.args[0].EvalJSON(b.ctx, row)
+	if err != nil {
+		return res, true, b.checkError(err)
+	}
+	if isNull {
+		return res, true, nil
+	}
+
+	for i := 1; i < len(b.args)-1; i += 2 {
+		// bad arguments should return null
+		// NOTE: mysql does not report error
+		// also they does not ignore it, they just return NULL
+		s, isNull, err := b.args[i].EvalString(b.ctx, row)
+		if isNull {
+			return res, true, b.checkError(json.ErrInvalidJSONPath.GenWithStackByArgs(s))
+		}
+		err = b.checkError(err)
+		if err != nil {
+			return res, true, err
+		}
+		pathExpr, err := json.ParseJSONPathExpr(s)
+		if err != nil {
+			return res, true, b.checkError(json.ErrInvalidJSONPath.GenWithStackByArgs(s))
+		}
+		if pathExpr.ContainsAnyAsterisk() {
+			return res, true, b.checkError(json.ErrInvalidJSONPathWildcard.GenWithStackByArgs(s))
+		}
+
+		var exists bool
+		obj, exists := res.Extract([]json.PathExpression{pathExpr})
+		if !exists {
+			// just do nothing for this
+			continue
+		}
+
+		if obj.TypeCode != json.TypeCodeArray {
+			// JSON_ARRAY_APPEND({"a": "b"}, "$", "c") => [{"a": "b"}, "c"]
+			// We should convert them to a single array first
+			obj = json.CreateBinary([]interface{}{obj})
+		}
+
+		var (
+			value  json.BinaryJSON
+			isnull bool
+		)
+
+		value, isnull, err = b.args[i+1].EvalJSON(b.ctx, row)
+		if err != nil {
+			return res, true, err
+		}
+
+		if isnull {
+			value = json.CreateBinary(nil)
```

Comment:
This is pretty tricky, according to [documentation](https://dev.mysql.com/doc/refman/5.7/en/json-modification-functions.html#function_json-array-append) of MySQL
> Returns NULL if any argument is NULL. 

but this is how it really behaves:
```
mysql> SELECT JSON_ARRAY_APPEND(@j, '$[1]', null);
+-------------------------------------+
| JSON_ARRAY_APPEND(@j, '$[1]', null) |
+-------------------------------------+
| ["a", ["b", "c", null], "d"]        |
+-------------------------------------+
1 row in set (0.00 sec)
```
@morgo any suggestion on the behavior here?

Reply:
I think it is better to follow the code than the documentation.